cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java

```diff
@@ -83,7 +90,66 @@ public AggregatedMetricsEmitter load(EmitterKey key) throws Exception {
   protected final void runOneIteration() throws Exception {
     final long timestamp = TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS);
     LOG.trace("Start log collection for timestamp {}", timestamp);
-    Iterator<MetricValue> metricsItor = getMetrics(timestamp);
+
+    final Iterator<MetricValue> rawMetricsItor = getMetrics(timestamp);
+    Iterator<MetricValue> metricsItor;
+    if (!recordMetaMetrics) {
+      metricsItor = rawMetricsItor;
+    } else {
+      // wrap the raw metrics iterator with an iterator that will publish meta metrics
+      metricsItor = new AbstractIterator<MetricValue>() {
```

Comment:
Move this into a class/method to make the logic here simple and straightforward.